cms/bundles/forms.py

```diff
@@ -318,55 +318,114 @@ def _sync_datasets_with_bundle_api(
+        if final_etag and final_etag != bundle.bundle_api_etag:
+            bundle.bundle_api_etag = final_etag
+            bundle.save(update_fields=["bundle_api_etag"])
+
+    @staticmethod
+    def _handle_api_add_items(
+        *,
+        client: BundleAPIClient,
+        bundle: "Bundle",
+        items: set["BundleDataset"],
+    ) -> str | None:
+        """Add or link content items to the bundle. Stops on first error. Advances ETag after each success."""
+        if not items:
+            return None
+
+        etag = None
+        for item in items:
+            content_item = build_content_item_for_dataset(item.dataset)
+            try:
+                # Client handles If-Match internally. We only read the returned ETag.
+                response = client.add_content_to_bundle(
+                    bundle_id=bundle.bundle_api_bundle_id,
+                    content_item=content_item,
+                )
                 etag = response["etag_header"]
 
-                # Extract content_id from the response
-                if content_id := extract_content_id_from_bundle_response(response, item.dataset):
-                    item.bundle_api_content_id = content_id
-                    item.save(update_fields=["bundle_api_content_id"])
-                    logger.info(
-                        "Added content %s to bundle %s in Bundle API with ID %s",
-                        item.dataset.namespace,
-                        bundle.pk,
-                        content_id,
+                content_id = extract_content_id_from_bundle_response(response, item.dataset)
+                if not content_id:
+                    logger.exception(
+                        "Could not find content_id in response for bundle",
+                        extra={"id": bundle.pk, "api_id": bundle.bundle_api_bundle_id},
                     )
-                else:
-                    logger.exception("Could not find content_id in response for bundle %s", bundle.pk)
+                    raise ValidationError("Bundle API did not return an ID for the added content")
 
-            if etag is not None:
-                bundle.bundle_api_etag = etag
-                bundle.save(update_fields=["bundle_api_etag"])
-        except BundleAPIClientError as e:
-            logger.exception("Failed to add content to bundle %s in Bundle API: %s", bundle.pk, e)
-            raise ValidationError("Could not communicate with the Bundle API") from e
+                item.bundle_api_content_id = content_id
+                item.save(update_fields=["bundle_api_content_id"])
+
+                logger.info(
+                    "Added content %s to bundle %s in Bundle API with ID %s",
+                    item.dataset.namespace,
+                    bundle.pk,
+                    content_id,
+                )
+            except BundleAPIClientError as e:
+                logger.exception("Failed to add content to bundle %s in Bundle API: %s", bundle.pk, e)
```

Comment:
NP: Since we are improving our logger calls by adding `extra`, perhaps we can update it here too?

Reply:
I will be following up with improvements to the error message as a whole. I'll address it in that PR if that's okay. That will involve surfacing API errors to users and making improvements to the messaging as well as tidying up to use extra.